File: src/main/java/com/jackmeng/stl/stl_RayCaster.java
// Copyright 2023 Jack Meng. All rights reserved.
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.

package com.jackmeng.stl;

import java.awt.image.*;
import java.awt.*;

public class stl_RayCaster
{
  private int screenWidth;
  private int screenHeight;
  private double[] playerPos;
  private double playerAngle;
  private double fov;
  private double[][] map;

  public stl_RayCaster(int screenWidth, int screenHeight, double[] playerPos, double playerAngle, double fov,
      double[][] map)
  {
    this.screenWidth = screenWidth;
    this.screenHeight = screenHeight;
    this.playerPos = playerPos;
    this.playerAngle = playerAngle;
    this.fov = fov;
    this.map = map;
  }

  public BufferedImage castRays()
  {
    BufferedImage image = new BufferedImage(screenWidth, screenHeight, BufferedImage.TYPE_INT_RGB);
    double angleStep = fov / screenWidth;
    double currentAngle = playerAngle - (fov / 2);
    for (int x = 0; x < screenWidth; x++)
    {
      double rayAngle = currentAngle * Math.PI / 180;
      double distanceToWall = 0;
      boolean hitWall = false;
      double[] rayPos = { playerPos[0], playerPos[1] };
      while (!hitWall && distanceToWall < map.length)
      {
        distanceToWall += 0.1;
        int testX = (int) (rayPos[0] + Math.sin(rayAngle) * distanceToWall);
        int testY = (int) (rayPos[1] + Math.cos(rayAngle) * distanceToWall);
        if (testX < 0 || testX >= map[0].length || testY < 0 || testY >= map.length)
        {
          hitWall = true;
          distanceToWall = map.length;
        }
        else if (map[testY][testX] == 1)
        {
          hitWall = true;
        }
      }
      assert distanceToWall != 0;
      int ceiling = (int) ((screenHeight / 2.0) - screenHeight / distanceToWall);
      int floor = screenHeight - ceiling;
      for (int y = 0; y < screenHeight; y++)
      {
        if (y < ceiling)
          image.setRGB(x, y, Color.BLUE.getRGB());
        else if (y > floor)
          image.setRGB(x, y, Color.GREEN.getRGB());
        else
          image.setRGB(x, y, Color.WHITE.getRGB());
      }
      currentAngle += angleStep;
    }
    return image;
  }

  public static void raycast(int[][] map, double posX, double posY, double dirX, double dirY, double planeX,
      double planeY, int[] pixels, int width, int height, Color color)
  {
    for (int x = 0; x < width; x++)
    {
      double cameraX = 2 * x / (double) width - 1;
      double rayDirX = dirX + planeX * cameraX;
      double rayDirY = dirY + planeY * cameraX;

      int mapX = (int) posX;
      int mapY = (int) posY;

      double sideDistX;
      double sideDistY;

      double deltaDistX = Math.abs(1 / rayDirX);
      double deltaDistY = Math.abs(1 / rayDirY);
      double perpWallDist;

      int stepX;
      int stepY;

      boolean hit = false;
      int side = 0;

      if (rayDirX < 0)
      {
        stepX = -1;
        sideDistX = (posX - mapX) * deltaDistX;
      }
      else
      {
        stepX = 1;
        sideDistX = (mapX + 1.0 - posX) * deltaDistX;
      }
      if (rayDirY < 0)
      {
        stepY = -1;
        sideDistY = (posY - mapY) * deltaDistY;
      }
      else
      {
        stepY = 1;
        sideDistY = (mapY + 1.0 - posY) * deltaDistY;
      }

      while (!hit)
      {
        if (sideDistX < sideDistY)
        {
          sideDistX += deltaDistX;
          mapX += stepX;
          side = 0;
        }
        else
        {
          sideDistY += deltaDistY;
          mapY += stepY;
          side = 1;
        }

        if (map[mapX][mapY] > 0)
          hit = true;
      }

      perpWallDist = side == 0 ? (mapX - posX + (1D - stepX) / 2) / rayDirX
          : (mapY - posY + (1D - stepY) / 2) / rayDirY;

      int lineHeight = (int) (height / perpWallDist);

      int drawStart = -lineHeight / 2 + height / 2;
      if (drawStart < 0)
        drawStart = 0;

      int drawEnd = lineHeight / 2 + height / 2;
      if (drawEnd >= height)
        drawEnd = height - 1;

      for (int y = drawStart; y <= drawEnd; y++)
      {
        int colorValue = color.getRGB();
        pixels[x + y * width] = colorValue;
      }
    }
  }
}

File: src/main/java/com/jackmeng/stl/stl_RayCaster.kt
// Copyright 2023 Jack Meng. All rights reserved.
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.
package com.jackmeng.stl

import java.awt.image.BufferedImage
import java.awt.Color

class stl_RayCaster(
		private val screenWidth:Int , private val screenHeight:Int , private val playerPos:DoubleArray , private val playerAngle:Double , private val fov:Double , private val map:Array<DoubleArray>
)
{
	fun castRays():BufferedImage
	{
		val image=BufferedImage(screenWidth , screenHeight , BufferedImage.TYPE_INT_RGB)
		val angleStep=fov/screenWidth
		var currentAngle=playerAngle-fov/2
		for (x in 0 until screenWidth)
		{
			val rayAngle=currentAngle*Math.PI/180
			var distanceToWall=0.0
			var hitWall=false
			val rayPos=doubleArrayOf(playerPos[0] , playerPos[1])
			while (!hitWall&&distanceToWall<map.size)
			{
				distanceToWall+=0.1
				val testX=(rayPos[0]+Math.sin(rayAngle)*distanceToWall).toInt()
				val testY=(rayPos[1]+Math.cos(rayAngle)*distanceToWall).toInt()
				if (testX<0||testX>=map[0].size||testY<0||testY>=map.size)
				{
					hitWall=true
					distanceToWall=map.size.toDouble()
				}
				else if (map[testY][testX]==1.0)
				{
					hitWall=true
				}
			}
			assert(distanceToWall!=0.0)
			val ceiling=(screenHeight/2.0-screenHeight/distanceToWall).toInt()
			val floor=screenHeight-ceiling
			for (y in 0 until screenHeight)
			{
				if (y<ceiling) image.setRGB(x , y , Color.BLUE.rgb)
				else if (y>floor) image.setRGB(
					x ,
					y ,
					Color.GREEN.rgb
				)
				else image.setRGB(x , y , Color.WHITE.rgb)
			}
			currentAngle+=angleStep
		}
		return image
	}
	
	companion object
	{
		fun raycast(
				map:Array<IntArray> , posX:Double , posY:Double , dirX:Double , dirY:Double , planeX:Double , planeY:Double , pixels:IntArray , width:Int , height:Int , color:Color
		)
		{
			for (x in 0 until width)
			{
				val cameraX=2*x/width.toDouble()-1
				val rayDirX=dirX+planeX*cameraX
				val rayDirY=dirY+planeY*cameraX
				var mapX=posX.toInt()
				var mapY=posY.toInt()
				var sideDistX:Double
				var sideDistY:Double
				val integral_Dist_X=0.1f
				val deltaDistX=Math.abs(integral_Dist_X/rayDirX)
				val deltaDistY=Math.abs(1/rayDirY)
				var perpWallDist:Double
				var stepX:Int
				var stepY:Int
				var hit=false
				var side=0
				if (rayDirX<0)
				{
					stepX=-1
					sideDistX=(posX-mapX)*deltaDistX
				}
				else
				{
					stepX=1
					sideDistX=(mapX+1.0-posX)*deltaDistX
				}
				if (rayDirY<0)
				{
					stepY=-1
					sideDistY=(posY-mapY)*deltaDistY
				}
				else
				{
					stepY=1
					sideDistY=(mapY+1.0-posY)*deltaDistY
				}
				while (!hit)
				{
					if (sideDistX<sideDistY)
					{
						sideDistX+=deltaDistX
						mapX+=stepX
						side=0
					}
					else
					{
						sideDistY+=deltaDistY
						mapY+=stepY
						side=1
					}
					if (map[mapX][mapY]>0) hit=true
				}
				perpWallDist=if (side==0) (mapX-posX+(1.0-stepX)/2)/rayDirX else (mapY-posY+(1.0-stepY)/2)/rayDirY
				val lineHeight=(height/perpWallDist).toInt()
				var drawStart=-lineHeight/2+height/2
				if (drawStart<0) drawStart=0
				var drawEnd=lineHeight/2+height/2
				if (drawEnd>=height) drawEnd=height-1
				for (y in drawStart..drawEnd)
				{
					val colorValue=color.rgb
					pixels[x+y*width]=colorValue
				}
			}
		}
	}
}